addons/enemybar/actionTracking.lua

```diff
@@ -0,0 +1,299 @@
+		        end
+		        
+		    -- Non-damaging spells
+		    elseif S{75,236,237,268,270,271}:contains(t.actions[1].message) then
+		        local effect = t.actions[1].param
+		        local spell = ai.param
+
+		        apply_debuff(target, effect, spell)
+		    end
+		end
+	end
+end
+
+function apply_debuff(target, effect, spell)	
+    if not tracked_debuff[target.id] then
+        tracked_debuff[target.id] = {}
+    end
+
+    local overwrites = res.spells[spell].overwrites or {}
+    if not did_overwrite(target, spell, overwrites) then
+        return
+    end
+
+    tracked_debuff[target.id][effect] = {target=target,spell=spell,effect=effect,time=os.time(),duration=res.spells[spell].duration or 0,pos={x=target.x,y=target.y}}
+end
+
+function did_overwrite(target, new, t)
+    if not tracked_debuff[target.id] then
+        return true
+    end
+    
+    for effect, tracked in pairs(tracked_debuff[target.id]) do
+        local old = res.spells[tracked.spell].overwrites or {}
+        
+        -- Check if there isn't a higher priority debuff active
+        if table.length(old) > 0 then
+            for _,v in ipairs(old) do
+                if new == v then
+                    return false
+                end
+            end
+        end
+        
+        -- Check if a lower priority debuff is being overwritten
+        if table.length(t) > 0 then
+            for _,v in ipairs(t) do
+                if tracked.spell == v then
+                    tracked_debuff[target.id][effect] = nil
+                end
+            end
+        end
+    end
+    return true
+end
+
+function check_conflicting_debuffs(ai)
+	local actor = windower.ffxi.get_mob_by_id(ai.actor_id)
+	if not actor then return end
+	if actor.is_npc then
+		-- the actor is an npc, let's check if they're supposed to be asleep/petrified/terror'd
+		local debuffs = tracked_debuff[actor.id]
+		if not debuffs then return end
+
+		for id,debuff in pairs(debuffs) do
+			if S{2,19,7,28}:contains(id) then
+				-- it was inactive, but now it's doing!
+				tracked_debuff[actor.id][id] = nil
+			end
+			if math.abs(debuff.pos.x-actor.x)>0.5 or math.abs(debuff.pos.y-actor.y)>0.5 then
+				-- it was locked in place, but now it's not!
+				tracked_debuff[actor.id][id] = nil
+			end
+		end
+	end
+end
+
+function clean_tracked_actions()
+	clean_actions_tick = clean_actions_tick - 1
+
+	if clean_actions_tick > 0 then return end
+
+	local player = windower.ffxi.get_mob_by_target("me")
+	local time = os.time()
+	for id,action in pairs(tracked_actions) do
+		-- for incomplete items, timeout at 30s.
+		if not action.complete and time - action.time > 30 then
+			tracked_actions[id] = nil
+
+		-- for complete actions, timeout at 3s.
+		elseif action.complete and time - action.time > 3 then
+			tracked_actions[id] = nil
+		end
+	end
+
+	for id,enmity in pairs(tracked_enmity) do
+		if time - enmity.time > 3 then
+			local mob = windower.ffxi.get_mob_by_id(enmity.mob.id)
```

Comment:
You call that function a few more times in here. These may be new IDs, so not sure if there's a way around it, just pointing it out. I won't make a mention of this again, just search for those occurrences and see what you find that you can optimize.